### PR TITLE
EventProbe: use redirected_tty_write BTF info instead of tty_write

### PR DESCRIPTION
**Pull request by mmat11**
Use a different function to obtain signature indexes as `tty_write` BTF info are missing on ARM systems. Read more: https://github.com/elastic/ebpf/pull/116#issue-1327583872

Co-Authored-By: Rhys Rustad-Elliott <rhys.rustad-elliott@elastic.co>

### CODE REVIEW

**Comment by rhysre** [comment]
Hmmm, mainline/5.19 is failing for arm64 and x86. You can ignore all failures except mainline (I'm working on those). But mainline was passing before and should keep passing.

I'm seeing the following:
```
[            ] stdout: Failed to unmarshal the following JSON: "": unexpected end of JSON input
```

Seems the testrunner is unable to unmarshal something that `EventsTrace` outputs. Digging into this further right now.

**Comment by mmat11** [comment]
> Hmmm, mainline/5.19 is failing for arm64 and x86. You can ignore all failures except mainline (I'm working on those). But mainline was passing before and should keep passing.
> 
> I'm seeing the following:
> 
> ```
> [            ] stdout: Failed to unmarshal the following JSON: "": unexpected end of JSON input
> ```
> 
> Seems the testrunner is unable to unmarshal something that `EventsTrace` outputs. Digging into this further right now.

missed that, will have a look later (probably something changed in tests, I am pretty sure it used to work)

**Review by nicholasberlin — APPROVED** [review]

File: non-GPL/Events/Lib/EbpfEvents.c [hunk 1]
```diff
@@ -233,6 +239,13 @@ static inline int probe_set_autoload(struct btf *btf, struct EventProbe_bpf *obj
         err = err ?: bpf_program__set_autoload(obj->progs.fexit__tcp_v6_connect, false);
     }
 
+    // tty_write is not present in all supported kernels' BTF info (eg. amazonlinux2 x86_64)
```
Comment (rhysre) [inline non-GPL/Events/Lib/EbpfEvents.c:242]: Why is this needed? It's true that `tty_write` isn't present in all supported kernel's BTF info, but that should simply be fixed by using `redirected_tty_write` as an analog. AFAICS we should never have to not load the probe.

**Comment by rhysre** [comment]
@mmat11 Figured out why this is failing on mainline/5.19.

Our "5.19" kernel is actually 6.0-rc1 (check the first line logged to the serial console):

```
Booting from ROM..ESCcESC[?7lESC[2J[    0.000000] Linux version 6.0.0-rc1 (root@lrishi-el) (x86_64-linux-gnu-gcc (Debian 10.2.1-6) 10.2.1 20210110, GNU ld (GNU Binutils for Debian) 2.35.2) #1 SMP PREEMPT_DYNAMIC Tue Aug 16 12:02:02 UTC 2022
```

6.0-rc1 added this commit: https://github.com/torvalds/linux/commit/fcb14cb1bdacec5b4374fe161e83fb8208164a85

`ii->iter_type` is 6 in the invocation of the `tty_write` probe on 6.0-rc1 (i.e. `ITER_UBUF`), so we need to be reading from `ii->ubuf` instead of `ii->iov->iov_base`.

TBH though an RC kernel should not be in our testing set, and this problem is unrelated to this PR, so I'd say we can fix that in a new PR.

**Comment by rhysre** [comment]
Spoke to @lrishi, we've removed the RC kernel and I've pushed a commit to use the new imagepack above.

File: non-GPL/Events/Lib/EbpfEvents.c (hunk 1 above)
Comment (mmat11) [reply non-GPL/Events/Lib/EbpfEvents.c:242]: we still want to hook `tty_write`, right? this block of code just loads a kprobe instead of the `tracing` counterpart when BTF info are missing

**Comment by mmat11** [comment]
> Our "5.19" kernel is actually 6.0-rc1 (check the first line logged to the serial console):

oh.. thanks for looking into it

File: non-GPL/Events/Lib/EbpfEvents.c [hunk 2]
```diff
@@ -194,15 +194,21 @@ static int probe_fill_relos(struct btf *btf, struct EventProbe_bpf *obj)
...
     err = err ?: FILL_FUNC_RET_IDX(obj, btf, vfs_rename);
 
-    // TODO: Re-enable tty_write probe when BTF issues are fixed
-#if 0
-    if (FILL_FUNC_ARG_EXISTS(obj, btf, tty_write, from)) {
-        err = err ?: FILL_FUNC_ARG_IDX(obj, btf, tty_write, buf);
-        err = err ?: FILL_FUNC_ARG_IDX(obj, btf, tty_write, count);
+    /* From https://github.com/elastic/ebpf/pull/116#issue-1327583872
+     *
+     * tty_write BTF info is not available on ARM systems with default
```
Comment (rhysre) [inline non-GPL/Events/Lib/EbpfEvents.c:199]: This isn't _quite_ right, it's not the default ARM64 config, it's when a (relocatable) ARM kernel is built with pahole < 1.22. Can you update this to say "tty_write BTF info is not available on ARM64 kernels built with pahole < 1.22 due to a bug in pahole".

File: non-GPL/Events/Lib/EbpfEvents.c (hunk 1 above)
Comment (rhysre) [reply non-GPL/Events/Lib/EbpfEvents.c:242]: Ok, I was a tad confused earlier. This code looks good to me. Can you clarify the comment though. I'd switch it to:

 "tty_write BTF information is not available on all supported kernels due to a pahole bug. If it is not present we can't attach a fentry/ program to it, so fallback to a kprobe"

**Review by rhysre — APPROVED** [review]
Two comment changes for clarity please, otherwise LGTM. 

Thanks for getting this done Mattia! 🚀
